pennylane/allocation.py

```diff
@@ -0,0 +1,275 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the commands for allocating and freeing wires dynamically.
+"""
+import uuid
+from typing import Optional, Sequence
+
+from pennylane.operation import Operator
+from pennylane.wires import Wires
+
+
+class DynamicWire:
+    """A wire whose concrete value will be determined later during a compilation step or execution.
+
+    Multiple dynamic wires can correspond to the same device wire as long as they are properly allocated and
+    freed.
+
+    Args:
+        key (Optional[str]): a ``uuid4`` string to uniquely identify the dynamic wire.
+    """
+
+    def __init__(self, key: Optional[uuid.UUID] = None):
+        if key:
+            self.key = key
+        else:
+            self.key = uuid.uuid4()
+
+    def __eq__(self, other):
+        if not isinstance(other, DynamicWire):
+            return False
+        return self.key == other.key
+
+    def __hash__(self):
+        return hash(("DynamicWire", self.key))
+
+    def __repr__(self):
+        return "<DynamicWire>"
+
+
+class Allocate(Operator):
+    """An instruction to values for dynamic wires.
+
+    Args:
+        wires (list[DynamicWire]): a list of dynamic wire values.
+
+    Keyword Args:
+        require_zeros (bool): Whether or not the wire must start in a ``0`` state.
+
+    ..see-also:: :func:`~.allocate`, :func:`~.allocate_ctx`.
+
+    """
+
+    def __init__(self, wires, require_zeros=True):
+        super().__init__(wires=wires)
+        self._hyperparameters = {
+            "require_zeros": require_zeros,
+        }
+
+    @property
+    def require_zeros(self):
+        """Whether or not the allocated wires are required to be in the zero state."""
+        return self.hyperparameters["require_zeros"]
+
+    @classmethod
+    def from_num_wires(cls, num_wires: int, require_zeros=True) -> "Allocate":
+        """Initialize an ``Allocate`` op from a number of wires instead of already constructed dynamic wires."""
+        wires = tuple(DynamicWire() for _ in range(num_wires))
+        return cls(wires=wires, require_zeros=require_zeros)
+
+
+class Deallocate(Operator):
+    """An instruction to deallocate the provided ``DynamicWire``'s.
+
+    Args:
+        wires (DynamicWire, Sequence[DynamicWire]): one or more dynamic wires to deallocate.
+        reset_to_original (bool): Whether or not the qubit will be in the same state upon being freed.
```

Comment:
I personally don't like "reset_to_original" because it sounds more like a request. Can we call it something like `"restored"` or `"uncomputed"`? @isaacdevlugt